renku/core/management/datasets.py

```diff
@@ -476,59 +476,72 @@ def is_protected_path(self, path):
         """Add a file or directory from a local filesystem."""
-        src = Path(os.path.abspath(path))
+        action = "symlink" if external else "copy"
+        absolute_dataset_data_dir = (self.path / get_dataset_data_dir(self, dataset)).resolve()
+        source_root = Path(get_absolute_path(path))
+        is_with_repo = is_subpath(path=source_root, base=self.path)
 
-        if not src.exists():
-            raise errors.ParameterError(f"Cannot find file/directory: {path}")
+        def check_recursive_addition(src: Path):
+            if src.resolve() == absolute_dataset_data_dir:
+                raise errors.ParameterError(f"Cannot recursively add path containing dataset's data directory: {path}")
 
-        dst = destination / src.name
+        def get_destination_root():
+            destination_exists = destination.exists()
+            destination_is_dir = destination.is_dir()
 
-        # if we have a directory, recurse
-        if src.is_dir():
-            if dst.exists() and not dst.is_dir():
-                raise errors.ParameterError(f'Cannot copy directory to a file: "{dst}"')
-            if src == (self.path / get_dataset_data_dir(self, dataset)).resolve():
-                raise errors.ParameterError(f"Cannot add dataset's data directory recursively: {path}")
+            if self.is_protected_path(source_root):
+                raise errors.ProtectedFiles([source_root])
 
-            if self.is_protected_path(src):
-                raise errors.ProtectedFiles([src])
+            check_recursive_addition(source_root)
 
-            files = []
-            for f in src.iterdir():
-                files.extend(
-                    self._add_from_local(dataset=dataset, path=os.path.abspath(f), external=external, destination=dst)
-                )
-            return files
-        else:
-            # Check if file is in the project and return it
-            path_in_repo = None
-            if is_external_file(path=src, client_path=self.path):
-                path_in_repo = path
-            else:
-                try:
-                    path_in_repo = src.relative_to(self.path)
-                except ValueError:
-                    pass
-                else:
-                    if self.is_protected_path(src):
-                        raise errors.ProtectedFiles([src])
+            if not source_root.exists():
+                raise errors.ParameterError(f"Cannot find source file: {path}")
+            if source_root.is_dir() and destination_exists and not destination_is_dir:
+                raise errors.ParameterError(f"Cannot copy directory '{path}' to non-directory '{destination}'")
 
-            if path_in_repo:
-                return [{"path": path_in_repo, "source": path_in_repo, "parent": self}]
+            name = source_root.name if destination_exists and destination_is_dir else ""
+            return destination / name
```

Comment:
```suggestion
            return destination / source_root.name if destination_exists and destination_is_dir else destination
```
Seems cleaner to me, composing a Path with an empty string seems weird to me.